### ISSUE TITLE
All endpoints and their dependencies are instantiated at the start of the application.

### ISSUE THREAD

**Issue by BrunoBeraud** [FastEndpoints/FastEndpoints#660]
Hello, when "app.UseFastEndpoints()" is called, it automatically discovers all the endpoints in the application and instantiates all of them with their dependencies and subdependencies.

The problem is that one of my endpoints uses a service that is registered with a factory in the IOC container.

```
 builder.Services.AddScoped<IMyService>(sp =>
 {

 var context = provider.GetRequiredService<IHttpContextAccessor>() ;
 ... // Depending on the http context, I'll create a specific service

 }) ;
```

I know I could use a different approach to achieve this (like having a dedicated service provider that would do this) but I think it's a shame that I can't use a more direct approach that would be the factory.

Also, isn't it penalizing, in terms of performance, to have this behavior?

I'd love to hear your opinion on this, thank you!

**Comment by dj-nitehawk** [comment]
what you can do is, create another implementation of `IMyService` that doesn't actually do any work and return it from the factory during startup. something like this:
```cs
bld.Services.AddScoped<IMyService>(
    sp =>
    {
        var ctx = sp.GetService<IHttpContextAccessor>()?.HttpContext;

        if (ctx is null)
            return new MyNullService();

        return new MyService(ctx.Connection.Id);
    });
```

> isn't it penalizing, in terms of performance, to have this behavior?

yes that is true but not by a huge margin. of course, that depends on how heavy the services are, in which case they'd be detecting startup and not do the work if it seems to be contributing too much of a delay to the startup sequence. but then again if a particular service is really expensive to initialize, and you expect that cost to be incurred per each request, then you have a different problem to deal with.

DI during startup cannot be avoided due to two reasons:
- it's not possible to run the `Configure()` method without instantiating the endpoint class (which requires constructor injection).
- people wanted the ability to use injected dependencies within the `Configure()` method.

basically, we made the choice to favor faster request processing over faster startup as well as the ability to use injected services in `Configure()`. this way, most services are materialized during app startup and ready to go when http requests come in (except for transient/scoped services of course).

bottom line is, FE startup is pretty fast, even with DI during startup.
if you are deploying to cloud functions and need the fastest possible cold starts, you should not be using FE, even though a lot of FE users do use it that way (probably because it's fast enough for their needs).

here's an example of a decent size project, which is not bad at all:

![screen-record-1713242553](https://github.com/FastEndpoints/FastEndpoints/assets/7043768/1a6eab96-6c76-40f4-861a-a77a58f1e178)
